### PR TITLE
added singleton level + COO test

### PR DESCRIPTION
**Pull request by bharath2438**
Added `singleton.hpp` to implement singleton level format and `singleton.cpp` having one test case with COO format using a compressed and a singleton level.

### CODE REVIEW

File: test/source/singleton_test.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+TEST_CASE("Singleton-COO")
+{
+    constexpr uintptr_t SIZE = 100;
+    constexpr uintptr_t SIZE1 = 100;
+    constexpr uint8_t ZERO = 0;
+
+    std::vector<uintptr_t> const pos{ 0, 7 };
+    std::vector<uintptr_t> const crd{ 0, 0, 1, 1, 3, 3, 3 };
+    std::vector<uintptr_t> const crd1{ 0, 1, 0, 1, 0, 3, 4 };
```
Comment (hameerabbasi) [inline test/source/singleton_test.cpp:17]:
```suggestion
    std::vector<uintptr_t> const crd0{ 0, 0, 1, 1, 3, 3, 3 };
    std::vector<uintptr_t> const crd1{ 0, 1, 0, 1, 0, 3, 4 };
```

**Review by hameerabbasi — CHANGES_REQUESTED** [review]
One minor change, just a variable rename.

**Comment by hameerabbasi** [comment]
You have to update it elsewhere was well.

**Comment by bharath2438** [comment]
> You have to update it elsewhere as well.

Yeah I have to do that but I made the suggested commit on my phone as I was in class, I'll do it as soon as I reach home.

File: test/source/singleton_test.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include <iostream>
+#include <vector>
+
+#include <xsparse/levels/compressed.hpp>
+#include <xsparse/levels/singleton.hpp>
+
+TEST_CASE("Singleton-COO")
+{
+    constexpr uintptr_t SIZE = 100;
+    constexpr uintptr_t SIZE1 = 100;
```
Comment (czgdp1807) [inline test/source/singleton_test.cpp:12]: Can we test on corner cases like SIZE = 100, SIZE1 = 0 or SIZE = 1, SIZE1 = 100? Not sure about the semantics of this test but my intent is to make sure that added code works fine with whatever corner cases are possible. Thanks.

Comment (bharath2438) [reply test/source/singleton_test.cpp:12]: > Can we test on corner cases like SIZE = 100, SIZE1 = 0 or SIZE = 1, SIZE1 = 100? Not sure about the semantics of this test but my intent is to make sure that added code works fine with whatever corner cases are possible. Thanks.

Yes that can be done, but as of now, SIZE by itself does not participate in the functions of singleton (pos_bounds and pos_access) so the value of SIZE should not affect the outcome of the test case. 
Maybe, I can add another test case having SIZE as 0 and SIZE1 as 100?

Comment (czgdp1807) [reply test/source/singleton_test.cpp:12]: Yes I would prefer to do so.

Comment (bharath2438) [reply test/source/singleton_test.cpp:12]: I added a test case for a 3rd order COO format which has a compressed level followed by two singleton levels. I've used 0, 100 and 1 as edge case sizes.

Comment (czgdp1807) [reply test/source/singleton_test.cpp:12]: Please feel free to resolve this thread.

**Comment by czgdp1807** [comment]
LGTM.

**Review by hameerabbasi — APPROVED** [review]
